Generate workflow xml file used for the OCN-ICE application

binli2337 opened the pull request:
**Description**

A new app choice "NG-GODAS" is added to the global-workflow for running the MOM6_CICE6_CDEPS_DATM application.

A new option "ATMTYPE" is added to the config.base.emc.dyn file. ATMTYPE is "DATA" if the atmosphere component is CDEPS. ATMTYPE is "MODEL" if the atmosphere component is FV3ATM.

A new switch "DO_POST" is added to the config.base.emc.dyn file. The GFS post-processing is performed only when DO_POST=true. For NG-GODAS application, no GFS post-processing is needed.

For the NG-GODAS application, there are only four sub-tasks in the workflow:

1. Get initial condition files for MOM6 and CICE6 and forcing files for CDEPS DATM.
2. Run the forecast model
3. Perform ocean/ice post-processing
4. Archive

Fixes: #838 

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Clone and Build tests on Hera
- [x] Forecast-only tests on Hera (both ATM and S2SW applications)
 
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

binli2337 committed:
updated the following files for NG-GODAS application:
parm/config/config.base.emc.dyn
parm/config/config.coupled_ic
ush/rocoto/setup_expt.py
ush/rocoto/setup_workflow_fcstonly.py

binli2337 committed:
Add parm/config/config.defaults.datm for the NG-GODAS application.

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

binli2337 committed:
Update parm/config/config.coupled_ic and sorc/link_workflow.sh.

binli2337 committed:
Merge branch 'add_NG_GODAS' of https://github.com/binli2337/global-workflow into add_NG_GODAS

binli2337 committed:
Update parm/config/config.coupled_ic and ush/rocoto/setup_workflow_fcstonly.py

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

WalterKolczynski-NOAA commented on parm/config/config.coupled_ic:
```diff
@@ -11,13 +11,16 @@ if [[ "$machine" == "WCOSS_DELL_P3" ]]; then
   export BASE_CPLIC="/gpfs/dell2/emc/modeling/noscrub/Walter.Kolczynski/global-workflow/IC"
 elif [[ "$machine" == "HERA" ]]; then
   export BASE_CPLIC="/scratch1/NCEPDEV/climate/role.ufscpara/IC"
+  export BASE_CPLIC2="/scratch1/NCEPDEV/nems/emc.nemspara/RT/DATM-MOM6-CICE6"
```
Can we get these added to the regular coupled IC location?

binli2337 replied:
@WalterKolczynski-NOAA Could you copy the datm forcing directory from /work/noaa/stmp/libin/CDEPS_DATM to /work/noaa/global/wkolczyn/noscrub/global-workflow/IC/CDEPS_DATM on Orion? Thanks!

WalterKolczynski-NOAA replied:
The version in role.ufscpara on Hera is the master version. I rsync to the other machines from there. Once that is updated, let me know and I will run the sync.

binli2337 replied:
@WalterKolczynski-NOAA The forcing files have been copied to /scratch1/NCEPDEV/climate/role.ufscpara/IC/CDEPS_DATM. Thanks.

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -185,6 +188,40 @@ case "${APP}" in
+    export flux_convergence=0.0
+    export flux_iteration=2
+    export flux_scheme=0
+# gefs forcing:
```
Comments should be indented to the same level as the code

binli2337 replied:
Done

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -185,6 +188,40 @@ case "${APP}" in
     export CCPP_SUITE="FV3_GFS_v17_coupled_p8"
     export confignamevarfornems="cpld${aero_nems_string}"
     ;;
+   NG-GODAS)
```
This indentation does not match the others.

binli2337 replied:
Done

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -185,6 +188,40 @@ case "${APP}" in
+    export ATM_NX_GLB=${ATM_NX_GLB:-"1536"}
+    export ATM_NY_GLB=${ATM_NY_GLB:-"768"}
+# cfsr forcing:
+    #export datm_src="cfsr"
```
Please remove the unused code

WalterKolczynski-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -64,6 +64,7 @@ for dir in fix_aer \
             fix_cice \
             fix_mom6 \
             fix_cpl \
+            fix_datm \
```
Does this need to be added to .gitignore?

binli2337 replied:
The existing fix/fix_* in .gitignore should work.

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -166,6 +167,8 @@ if [[ $DO_AERO = "YES" ]]; then
 else
     aero_nems_string=""
 fi
+export DO_ATM="YES"
```
Since DATM basically replaces ATM, I wonder if instead of DO_ATM this should be ATM="Model" or "Data"?

aerorahul replied:
Since `DO_` flags are either `YES|NO`, I suggest, we stick with that.

aerorahul replied:
If another variable is needed for `ATMTYPE`, then we can discuss other options

WalterKolczynski-NOAA replied:
The point was to use ATMTYPE instead of DO_ATM.

binli2337 replied:
Done

WalterKolczynski-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -240,13 +240,13 @@ def input_args():
     # forecast only mode additional arguments
     forecasts.add_argument('--app', help='UFS application', type=str, choices=[
-        'ATM', 'ATMW', 'S2S', 'S2SW'], required=False, default='ATM')
+        'ATM', 'ATMW', 'S2S', 'S2SW','NG-GODAS'], required=False, default='ATM')
```
Comma should have a space after it.

KateFriedman-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -404,13 +436,22 @@ def get_workflow(dict_configs, cdump='gdas'):
-    dep_dict = {'type':'data', 'data':data}
-    deps.append(rocoto.add_dependency(dep_dict))
-    dependencies = rocoto.create_dependency(dep_condition='or', dep=deps)
+    if do_atm in ['Y', 'YES']:
```
I second Walter's comment above about setting ATM="Model" or "Data". This will need to be updated as well.

binli2337 replied:
Done

KateFriedman-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -543,10 +585,16 @@ def get_workflow(dict_configs, cdump='gdas'):
-        dep_dict = {'type': 'data', 'data': data}
-        deps.append(rocoto.add_dependency(dep_dict))
-        dependencies = rocoto.create_dependency(dep=deps)
+        if do_datm in ['NO']:
```
Update this based on Walter's comment above about ATM="Model" or "Data".

KateFriedman-NOAA reviewed (commented):
Concur with Walter's comments and change requests. Will do my review after those are addressed.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -362,8 +399,8 @@ export netcdf_diag=".true."
-export DO_METP="YES"         # Run METPLUS jobs - set METPLUS settings in config.metp
-export DO_VSDB="YES"         # Run VSDB package - set VSDB settings in config.vrfy
+export DO_METP=${DO_METP:-"YES"}        # Run METPLUS jobs - set METPLUS settings in config.metp
+export DO_VSDB=${DO_VSDB:-"YES"}         # Run VSDB package - set VSDB settings in config.vrfy
```
Please move this section before the `case "${APP}"` block and remove the need for variables.
e.g.
```
export DO_METP="YES"
```

binli2337 replied:
Done.

aerorahul reviewed (changes requested):
Please see comments.

binli2337 committed:
Update the following four files:
 parm/config/config.base.emc.dyn
 parm/config/config.coupled_ic
 ush/rocoto/setup_expt.py
 ush/rocoto/setup_workflow_fcstonly.py

binli2337 committed:
Add ATMTYPE and remove DO_ATM.

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

WalterKolczynski-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -240,6 +240,8 @@ def get_workflow(dict_configs, cdump='gdas'):
     machine = base.get('machine', wfu.detectMachine())
     hpssarch = base.get('HPSSARCH', 'NO').upper()
     app = base.get('APP', "ATM").upper()
+    atmtype = base.get('ATMTYPE', 'MODEL').upper()
```
This is a minor thing, but the python variable should probably be atm_type.

WalterKolczynski-NOAA reviewed (changes requested):
One extremely minor thing.

binli2337 committed:
Update setup_workflow_fcstonly.py to use atm_type.

WalterKolczynski-NOAA reviewed (approved):
Update the PR to reflect the most recent changes and I think we're good. It should also describe the addition of the new app choice.

binli2337 committed:
Merge branch 'develop' into add_NG_GODAS

WalterKolczynski-NOAA commented on parm/config/config.defaults.datm:
```diff
@@ -0,0 +1,21 @@
+#!/bin/ksh -x
```
Please make this bash

WalterKolczynski-NOAA commented on parm/config/config.base.emc.dyn:
```diff
@@ -184,6 +190,28 @@ case "${APP}" in
+    export DO_METP="NO"
+    export DO_VRFY="NO"
+    export RUN="data"
+    export flux_convergence=0.0
```
All these model variables should be moved to the `config.defaults.datm`

binli2337 replied:
Done.

WalterKolczynski-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -296,6 +299,33 @@ def get_workflow(dict_configs, cdump='gdas'):
         tasks.append(task)
         tasks.append('\n')
 
+    elif app in ['NG-GODAS']:
```
Please consolidate with S2S applications

binli2337 replied:
Done. Thanks.

WalterKolczynski-NOAA reviewed (changes requested):
Minor changes requested jointly with Rahul and Kate

binli2337 committed:
* update  parm/config/config.base.emc.dyn
* update  parm/config/config.defaults.datm
* update  ush/rocoto/setup_workflow_fcstonly.py

WalterKolczynski-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -260,14 +260,21 @@ def get_workflow(dict_configs, cdump='gdas'):
+        if atm_type in ["MODEL"]:
+            deps = []
+            # ATM ICs
+            #base_cplic = dict_configs['coupled_ic']['BASE_CPLIC']
```
Remove this now redundant line, and the other identical line below

binli2337 replied:
Done.

WalterKolczynski-NOAA reviewed (changes requested):
Almost there

binli2337 committed:
Update setup_workflow_fcstonly.py.

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

aerorahul reviewed (approved):
Looks good.
I could ask why some variables are `export` and others not in `config.defaults.datm`.  That would get addressed when the workflow is setup and executed.

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

guillaumevernieres reviewed (commented):
Thanks for doing this @binli2337 . I have next to 0 expertise on the global workflow, but I would like to suggest dropping the `NG-GODAS` name to something that follows the "convention" for the other short keys and attempt to describe the model configuration via a few letters. How about `OCN-ICE`? Or `NOATM`?

binli2337 commented:
@guillaumevernieres I will change NG-GODAS to OCN-ICE. Thanks for the suggestion.

binli2337 committed:
Change application name from NG-GODAS to OCN-ICE.

aerorahul commented:
If only it was that simple. 
We use the App name from the Model. So, unless we change it in the model, it is stuck here.

guillaumevernieres commented:
> If only it was that simple. We use the App name from the Model. So, unless we change it in the model, it is stuck here.

No worries @binli2337 and @aerorahul . It's just that it is weird name for a model ... _I think I can live with it!_

binli2337 committed:
Merge branch 'develop' into add_NG_GODAS

guillaumevernieres reviewed (approved):
Adding my review as well. I have no comments on the code, but I did test @binli2337 's added feature in this branch and it does what this PR description says it should do, so :+1:

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

binli2337 committed:
Merge branch 'NOAA-EMC:develop' into add_NG_GODAS

WalterKolczynski-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -236,11 +236,11 @@ def input_args():
     # forecast only mode additional arguments
     forecasts.add_argument('--app', help='UFS application', type=str, choices=[
-        'ATM', 'ATMA', 'ATMW', 'S2S', 'S2SW', 'S2SWA'], required=False, default='ATM')
+        'ATM', 'ATMA', 'ATMW', 'S2S', 'S2SW', 'S2SWA' 'OCN-ICE'], required=False, default='ATM')
```
Missing a comma after S2SWA.

WalterKolczynski-NOAA commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -253,18 +255,24 @@ def get_workflow(dict_configs, cdump='gdas'):
-    if 'S2S' in app:
-        # Copy prototype ICs
-        deps = []
+    if app in ['S2S', 'S2SW', 'OCN-ICE']:
```
This won't work the same as before (no S2SWA). Please change to `if 'S2S' in app or app in ['OCN-ICE']`

WalterKolczynski-NOAA reviewed (changes requested):
Two things upon reviewing the code again.

This should be tested with ATM-only 2½-cycle cycled mode and a one-cycle S2SWA coupled run.

binli2337 committed:
Updated setup_expt.py and setup_workflow_fcstonly.py.

binli2337 commented:
The one-cycle S2SW coupled test is successful.
